exercises-solutions/src/main/scala/Ch6.scala

```diff
@@ -1,13 +1,519 @@
+      type J[A] = A
+
+      implicit val functorOptJ = new Functor[Lambda[X => Option[J[X]]]] {
+        override def map[A, B](optJA: Option[J[A]])(f: A => B): Option[J[B]] = optJA match {
+          case Some(a: J[A]) => Some(f(a))
+          case _ => None
+        }
+      }
+
+      implicit val filterableOptJ = new Filterable[Lambda[X => Option[J[X]]]] {
+        // 1 + J[1+A] => 1 + J
+        // Note: Option[J[Option[A]]] = Option[Option[A]] and Option[J[A]] = Option[A]
+        override def deflate[A](optJOptA: Option[J[Option[A]]]): Option[J[A]] =
+          optJOptA.flatten 
+      }
+
+      // Now we implement filterable functor F[J, J, A]
+      implicit val functorF = new Functor[Lambda[X => F[J, J, X]]] {
+        override def map[A, B](fa: F[J, J, A])(f: A => B): F[J, J, B] = fa match {
+          case FG(ga) => FG(ga.map(f))
+          case FH(ha) => FH(ha.map(f))
+          case F1() => F1()
+        }
+      }
+
+      implicit val filterableF = new Filterable[Lambda[X => F[J, J, X]]] {
+        override def deflate[A](fOptA: F[J, J, Option[A]]): F[J, J, A] = fOptA match {
+          case FG(gOptA: J[Option[A]]) => {
+            val optJOptA: Option[J[Option[A]]] = Some(gOptA)
+            val optJA: Option[J[A]] = optJOptA.deflate
+            optJA match {
+              case Some(a: J[A]) => FG(a)
+              case None => F1()
+            }
+          }
+          case FH(hOptA: J[Option[A]]) => {
+            val optJOptA: Option[J[Option[A]]] = Some(hOptA)
+            optJOptA.deflate match {
+              case Some(a: J[A]) => FH(a)
+              case None => F1()
+            }
+          }
+          case _ => F1()
+        }
+      }
+    }
+ 
+    object Problem6 {
+      // F[A] = A + (Int => A)
+      // Show F[_] is not filterable
+
+      // type F[A] = Either[A, Int => A]
+
+      // //implicit val functorF =
+      // implicit val filterableF = new Filterable[F] {
+      //   override def deflate[A](fOptA: F[Option[A]]): F[A] = fOptA match {
+      //     case Left(Some(a: A)) => Left(a)
+      //     case Left(None) => ??? // there is no way to produce a value of type A or Int => A
+      // }
+    }
+
+    object Problem7 {
+      // Given contrafunctor C[A], show that we can define "deflate: C[1 + A] => C[A]"
+      // Solution: just wrap Some(...) to the value before the arrow, and use that as
+      // argument into C[Option[A]]
+
+      type C[A] = Option[A] => Int
+
+      implicit val contrafunctorC = new Contravariant[C] {
+        override def contramap[A, B](ca: C[A])(f: B ⇒ A): C[B] = {
+          case Some(b: B) => ca(Some(f(b)))
+          case None => ca(None)
+        }
+      }
+
+      implicit val contrafilterableC = new ContraFilterable[C] {
+        override def inflate[A](ca: C[A]): C[Option[A]] = {
+          case Some(optA: Option[A]) => ca(optA)
+          case None => ca(None)
+        }
+      }
+
+      def deflateC[A](cOptA: C[Option[A]]): C[A] = { (optA: Option[A]) =>
```

Comment:
Instead, we need to show that `deflate` can be defined for a contrafunctor.

So, we want a PTVF with this type signature:

```scala
def deflate[C[_]: Contravariant, A]: C[Option[A]] => C[A] = { ... }
```